### ISSUE TITLE
hashlib.algorithms_available doesn't account for backends loaded dynamically at runtime

### ISSUE THREAD

**Issue by mgorny** [python/cpython#92874]
**Bug report**

If additional OpenSSL backends are loaded at runtime (this happens e.g. when importing [PyOpenSSL](https://www.pyopenssl.org/en/stable/)), the `hashlib.algorithms_available` doesn't get updated for the newly available hashes.

Trivial reproducer:
```python
>>> import ctypes, hashlib
>>> ctypes.CDLL("libssl.so").OSSL_PROVIDER_load(None, b"legacy")
-1311283920
>>> "md4" in hashlib.algorithms_available
False
>>> hashlib.new("md4")
<md4 _hashlib.HASH object @ 0x7f404308faf0>
```

**Your environment**

- CPython versions tested on: 3.11.0b1, 3.10.4 with 1b6acaad9a18b2498386c60f24351ab749061e3a backported, 3.9.12
- Operating system and architecture: Gentoo Linux amd64
- OpenSSL version: 3.0.3

**Comment by jakirkham** [comment]
This looks very similar to issue ( https://github.com/python/cpython/issues/92876 ). Should one of these be closed? Or am I missing some difference between these?

**Comment by mgorny** [comment]
No, not really. Solving #92876 would somewhat avoid one particular example but the issue would persist if one of the libraries loaded yet another provider.

**Comment by picnixz** [comment]
One way to mitigate this is to add a function that can modify the `hashlib.algorithms_available` set. It would be too complicate to make it automatic. Would it help solving your issue?

**Comment by mgorny** [comment]
I don't think it's really worth the effort. I wouldn't really call it an issue, rather an observed inconsistency.

A simpler way to avoid it would be to make `hashlib.new()` reject algorithms not in `algorithms_available`, but that probably has more downsides than real benefits.
